test/jdk/tools/sincechecker/SinceChecker.java

```diff
@@ -0,0 +1,769 @@
+                "method: javax.lang.model.element.RecordComponentElement javax.lang.model.util.Elements.recordComponentFor(javax.lang.model.element.ExecutableElement)",
+                "method: java.util.List javax.lang.model.util.ElementFilter.recordComponentsIn(java.lang.Iterable)",
+                "method: java.util.Set javax.lang.model.util.ElementFilter.recordComponentsIn(java.util.Set)",
+                "method: java.util.List javax.lang.model.element.TypeElement.getRecordComponents()",
+                "field: javax.lang.model.element.ElementKind:RECORD",
+                "field: javax.lang.model.element.ElementKind:RECORD_COMPONENT",
+                "field: javax.lang.model.element.ElementKind:BINDING_VARIABLE",
+                "field: com.sun.source.tree.Tree.Kind:RECORD",
+                "field: sun.reflect.annotation.TypeAnnotation.TypeAnnotationTarget:RECORD_COMPONENT",
+                "class: java.lang.reflect.RecordComponent",
+                "class: java.lang.runtime.ObjectMethods",
+                "field: java.lang.annotation.ElementType:RECORD_COMPONENT",
+                "class: java.lang.Record",
+                "method: boolean java.lang.Class.isRecord()",
+                "method: java.lang.reflect.RecordComponent[] java.lang.Class.getRecordComponents()",
+                "field: javax.lang.model.element.Modifier:SEALED",
+                "field: javax.lang.model.element.Modifier:NON_SEALED",
+                "method: javax.lang.model.element.TypeElement:getPermittedSubclasses:()",
+                "method: java.util.List com.sun.source.tree.ClassTree.getPermitsClause()",
+                "method: boolean java.lang.Class.isSealed()",
+                "method: java.lang.constant.ClassDesc[] java.lang.Class.permittedSubclasses()"
+        ));
+
+        LEGACY_PREVIEW_METHODS.put("16", Set.of(
+                "field: jdk.jshell.Snippet.SubKind:RECORD_SUBKIND",
+                "field: javax.lang.model.element.Modifier:SEALED",
+                "field: javax.lang.model.element.Modifier:NON_SEALED",
+                "method: javax.lang.model.element.TypeElement:getPermittedSubclasses:()",
+                "method: java.util.List com.sun.source.tree.ClassTree.getPermitsClause()",
+                "method: boolean java.lang.Class.isSealed()",
+                "method: java.lang.constant.ClassDesc[] java.lang.Class.permittedSubclasses()"
+        ));
+
+        // java.lang.foreign existed since JDK 19 and wasn't annotated - went out of preview in JDK 22
+        LEGACY_PREVIEW_METHODS.put("19", Set.of(
+                "package: java.lang.foreign"
+        ));
+        LEGACY_PREVIEW_METHODS.put("20", Set.of(
+                "package: java.lang.foreign"
+        ));
+        LEGACY_PREVIEW_METHODS.put("21", Set.of(
+                "package: java.lang.foreign"
+        ));
+    }
+
+    private final class EffectiveSourceSinceHelper implements AutoCloseable {
```

Comment:
Consider putting a doc comment describing the purpose of this class. The name  hints at something useful, but is not specific enough by itself.

Reply:
Added it, thanks